SIMPLE LINEAR REGRESSION: y = intercept + slope(x) + errorterm
- Linear Regression function parameter:
    1. X:np.array -> [1,2,3] "One-Dimension Array"
    2. y:np.array  -> [4,5,6] 'One-Dimension Array"
    * returns -> (slope, intercept)
- Predictions function parameter:
    1. slope: float
    2. intercept: float
    3. X: np.array -> [1,2,3]
    * returns -> predictions:np.array
- R2 function:
    1. y_true: np.array -> [1,2,3]
    2. predictions: np.array -> [1,2,3]
    * returns -> R2:float



In [203]:
import numpy as np

def linear_regression(X:np.array, y:np.array)->float:
 
    # Calculate the mean of X and y
    mean_X = np.mean(X)
    mean_y = np.mean(y)

    # Calculate the slope (m)
    numerator = np.sum((X - mean_X) * (y - mean_y))
    denominator = np.sum((X - mean_X) ** 2)
    slope = numerator / denominator

    # Calculate the intercept (b)
    intercept = mean_y - slope * mean_X

    return slope, intercept


In [204]:
num_friends = [100.0,49,41,40,25,21,21,19,19,18,18,16,15,15,15,15,14,14,13,13,13,13,12,12,11,10,10,10,10,10,10,10,10,10,10,10,10,10,10,10,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,9,8,8,8,8,8,8,8,8,8,8,8,8,8,7,7,7,7,7,7,7,7,7,7,7,7,7,7,7,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,6,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,5,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]
daily_minutes = [1,68.77,51.25,52.08,38.36,44.54,57.13,51.4,41.42,31.22,34.76,54.01,38.79,47.59,49.1,27.66,41.03,36.73,48.65,28.12,46.62,35.57,32.98,35,26.07,23.77,39.73,40.57,31.65,31.21,36.32,20.45,21.93,26.02,27.34,23.49,46.94,30.5,33.8,24.23,21.4,27.94,32.24,40.57,25.07,19.42,22.39,18.42,46.96,23.72,26.41,26.97,36.76,40.32,35.02,29.47,30.2,31,38.11,38.18,36.31,21.03,30.86,36.07,28.66,29.08,37.28,15.28,24.17,22.31,30.17,25.53,19.85,35.37,44.6,17.23,13.47,26.33,35.02,32.09,24.81,19.33,28.77,24.26,31.98,25.73,24.86,16.28,34.51,15.23,39.72,40.8,26.06,35.76,34.76,16.13,44.04,18.03,19.65,32.62,35.59,39.43,14.18,35.24,40.13,41.82,35.45,36.07,43.67,24.61,20.9,21.9,18.79,27.61,27.21,26.61,29.77,20.59,27.53,13.82,33.2,25,33.1,36.65,18.63,14.87,22.2,36.81,25.53,24.62,26.25,18.21,28.08,19.42,29.79,32.8,35.99,28.32,27.79,35.88,29.06,36.28,14.1,36.63,37.49,26.9,18.58,38.48,24.48,18.95,33.55,14.24,29.04,32.51,25.63,22.22,19,32.73,15.16,13.9,27.2,32.01,29.27,33,13.74,20.42,27.32,18.23,35.35,28.48,9.08,24.62,20.12,35.26,19.92,31.02,16.49,12.16,30.7,31.22,34.65,13.13,27.51,33.2,31.57,14.1,33.42,17.44,10.12,24.42,9.82,23.39,30.93,15.03,21.67,31.09,33.29,22.61,26.89,23.48,8.38,27.81,32.35,23.84]
outlier = num_friends.index(100)    # index of outlier

num_friends_good = [x
                    for i, x in enumerate(num_friends)
                    if i != outlier]

daily_minutes_good = [x
                      for i, x in enumerate(daily_minutes)
                      if i != outlier]
slope, intercept = linear_regression(num_friends_good, daily_minutes_good)
print("Slope :",slope)
print("Intercept :", intercept)

Slope : 0.9038659456058723
Intercept : 22.947552413468976


- we expect user with n friends to spend 22.95 + n*0.903 mins on the site each day. 
- we predict that a user with no friends would stiill spend about 23 mins a day on site. and for each additional friend, we expect a user to spend almost a minute more on site.
- we need R2 -> coef of determination to measure if we have fit the data, measures the fraction of total variation in the dependent variable 

In [205]:
#function to make prediction
def predict(intercept: float, slope:float, x_i:np.array) -> np.array:
    y_pred = (slope *x_i) + intercept
    return y_pred

predictions = predict(intercept, slope, np.array(num_friends_good))

In [206]:
#compute R2 - coef of determination = 1- (SSR/SST)
def compute_r2(y_true, y_pred):
    residual = y_true - y_pred
    mean_y_true = np.mean(y_true)
    #** is power of
    #total sum of square
    total_variance = np.sum((y_true - mean_y_true) ** 2)
    #sum of squared errors = SSR
    explained_variance = np.sum((residual) ** 2)

    r2 = 1- (explained_variance / total_variance)

    return r2

r2_value = compute_r2(np.array(daily_minutes_good), predictions)

print(f"R-squared: {r2_value}")


R-squared: 0.32910783778362984


higher the r2 is better model fits the data. now our r2 is 0.329 indicates that our model is only okay at fitting the data, clearly there are other factors at play.

Simple Linear Regression (from other author)
- TESTING WITH OTHER DATASET

In [217]:
#IMPORT DATA USING NUMPY 
import numpy as np 
# using loadtxt()
Data = np.loadtxt("Salary_Data.csv",
                 delimiter=",", dtype=str)

#display(Data)
#print(Data.ndim)
#print(Data.shape)
#print(Data.size)

In [218]:
#Process Data
Data_without_first_row = np.delete(Data, 0, axis=0)
shuffled_data = np.random.permutation(Data_without_first_row)
#print("Original 2D array:")
#print("Shuffled copy:")
#print(shuffled_data)

In [209]:
#training and testing set size
train_size=int(0.75*np.size(shuffled_data,0))
test_size=int(0.25*np.size(shuffled_data,0))
print("Training set size : "+ str(train_size))
print("Testing set size : "+str(test_size))

Training set size : 22
Testing set size : 7


In [210]:
#SPLIT FEATURES AND PREDICTOR
#Getting features from dataset
X=shuffled_data[:,[1]]
y=shuffled_data[:,0]


In [211]:
#SPLIT TRAIN/TEST
#training set split
X_train=X[0:train_size,:]
X_train = X_train.flatten()
X_train = X_train.astype(np.float_)

y_train=y[0:train_size]
y_train = y_train.flatten()
y_train = y_train.astype(np.float_)

#testing set split
X_test=X[train_size:,:]
X_test = X_test.flatten()
X_test = X_test.astype(np.float_)
Y_test=y[train_size:]
Y_test = Y_test.flatten()
Y_test = Y_test.astype(np.float_)

In [212]:
#Find slope, intercept, prediction and R2 for TRAINING data
slope_train, intercept_train = linear_regression(X_train, y_train)
predictions_train = predict(intercept_train, slope_train, np.array(X_train))
r2_train= compute_r2(np.array(y_train),predictions_train)
print("Slope :",slope_train)
print("Intercept :", intercept_train)
print("R-squared :", r2_train)

Slope : 0.00010212942351765587
Intercept : -2.378410713019366
R-squared : 0.9638606804844896


In [213]:
#Find slope and intercept for TESTING SET
slope_test, intercept_test = linear_regression(X_test, Y_test)
predictions_test = predict(intercept_test,slope_test, np.array(X_test))
r2_test = compute_r2(np.array(Y_test),predictions_test)
print("Slope :",slope_test)
print("Intercept :", intercept_test)
print("R-Squared :", r2_test)

Slope : 9.094873067726321e-05
Intercept : -1.8946963640419776
R-Squared : 0.9215724756896623
